src/Markup/Avalonia.Markup.Xaml/XamlIl/Runtime/XamlIlRuntimeHelpers.cs

```diff
@@ -25,7 +25,11 @@ public static class XamlIlRuntimeHelpers
                 var scope = parentScope != null ? new ChildNameScope(parentScope) : (INameScope)new NameScope();
                 var obj = builder(new DeferredParentServiceProvider(sp, resourceNodes, rootObject, scope));
                 scope.Complete();
-                return new ControlTemplateResult((IControl)obj, scope);
+                if (obj is IControl control)
+                {
+                    return new TemplateResult<IControl>(control, scope);
+                }
+                return new TemplateResult<object>(obj, scope);
```

Comment:
That's the only concerning part here.

Reply:
Not sure how to solve this better

Reply:
We might want to introduce a generic transformation factory. 
We can probably merge it as is, since we can always make TemplateResult<T> to be co-variant and keep the compatibility.